### PR TITLE
Improve RoomList render time when filtering

### PR DESCRIPTION
**Pull request by gsouquet**
Improves vector-im/element-web#14750
God speed unleashed when used with vector-im/element-web#16969

Benchmark below
* Figure 1: After my improvements
* Figure 2: Before my improvements

Screen Shot 2021-04-15 at 16 09 37

* A few events were not cleaned up properly
* Some work is done is the constructor when it could be offloaded at a later stage
* Mounting and unmounting `RoomSublist` is quite expensive. It ends up being much cheaper to hide rather than unmounting it when empty. Trading a tiny bit of ram for faster filtering

### CODE REVIEW

File: src/components/views/rooms/RoomList.tsx [hunk 1]
```diff
@@ -289,12 +285,11 @@ export default class RoomList extends React.PureComponent<IProps, IState> {
...
         this.tagAesthetics = objectShallowClone(TAG_AESTHETICS);
         this.updateDmAddRoomAction();
-
-        this.dispatcherRef = defaultDispatcher.register(this.onAction);
-        this.roomStoreToken = RoomViewStore.addListener(this.onRoomViewStoreUpdate);
     }
 
     public componentDidMount(): void {
+        this.dispatcherRef = defaultDispatcher.register(this.onAction);
+        this.roomStoreToken = RoomViewStore.addListener(this.onRoomViewStoreUpdate);
```
Comment (turt2live) [inline src/components/views/rooms/RoomList.tsx:292]: is it actually more performant to move things to `componentDidMount`? Over the last couple years we've started using the constructor more and more, slowly getting rid of `componentDidMount` - putting aside arguments about whether that is correct behaviour, I'm curious if there's a performance impact.

File: src/components/views/rooms/RoomSublist.tsx [hunk 2]
```diff
@@ -60,6 +60,11 @@ export const HEADER_HEIGHT = 32; // As defined by CSS
 
 const MAX_PADDING_HEIGHT = SHOW_N_BUTTON_HEIGHT + RESIZE_HANDLE_HEIGHT;
 
+const ALWAYS_VISIBLE_TAGS: TagID[] = [
+    DefaultTagID.DM,
+    DefaultTagID.Untagged,
+];
+
```
Comment (turt2live) [inline src/components/views/rooms/RoomSublist.tsx:67]: this feels like the sort of thing that should still be in the RoomList component (or appropriate model once we get there) - we can pass a `forceVisible` prop though, or if we can word it without using the word "force" then even better.

**Review by turt2live — COMMENTED** [review]
It also looks like there's failing end to end tests, which look like legitimate failures

**Comment by turt2live** [comment]
I'm not sure how to interpret the screenshot: it looks like we're spending *more* time in code?

Overall this looks sensible, but I am curious how it interacts with the tab indexing/accessibility. It'd also be good to get a comment somewhere in the code to say we deliberately keep the sublists rendered so we don't accidentally optimize it back out 😅

File: src/components/views/rooms/RoomList.tsx (hunk 1 above)
Comment (gsouquet) [reply src/components/views/rooms/RoomList.tsx:292]: It depends, generally speaking there can be small performance improvements when the component is rendered but not mounted 

But in my opinion the real power of this is that you colocate all your event listeners, if you have some event listeners that you need to place on DOM elements, you will have to perform that in `componentDidMount`. And on top of that if it is immediatly followed by `componentWillUnmount` you're one step closer to remembering to clean up your listeners

It also helps in tests where you shallow render your components

Overall moving away from lifecycles makes sense, this is pretty much what the React team has been advocating for, but their solution seems to be using hooks instead

File: src/components/views/rooms/RoomSublist.tsx (hunk 2 above)
Comment (gsouquet) [reply src/components/views/rooms/RoomSublist.tsx:67]: > I'm not sure how to interpret the screenshot: it looks like we're spending _more_ time in code?

I've read my description again, and it is utterly confusing, the left figure is AFTER my improvements. and the right figure is BEFORE 👀 

Jonathan (jboi) also took a recording and there, the left figure is BEFORE and the right figure is AFTER (sorry this is all very confusing)
Screen Shot 2021-04-15 at 19 49 20

> Overall this looks sensible, but I am curious how it interacts with the tab indexing/accessibility.

Generally speaking screen readers will ignore everything that is `display: none;`

> It'd also be good to get a comment somewhere in the code to say we deliberately keep the sublists rendered so we don't accidentally optimize it back out 😅

Will do 👍

**Comment by gsouquet** [comment]
> I'm not sure how to interpret the screenshot: it looks like we're spending _more_ time in code?

I've read my description again, and it is utterly confusing, the left figure is AFTER my improvements. and the right figure is BEFORE 👀 

Jonathan (jboi) also took a recording and there, the left figure is BEFORE and the right figure is AFTER (sorry this is all very confusing)
Screen Shot 2021-04-15 at 19 49 20

> Overall this looks sensible, but I am curious how it interacts with the tab indexing/accessibility.

Generally speaking screen readers will ignore everything that is `display: none;`

> It'd also be good to get a comment somewhere in the code to say we deliberately keep the sublists rendered so we don't accidentally optimize it back out 😅

Will do 👍

File: test/end-to-end-tests/src/usecases/signup.js [hunk 3]
```diff
@@ -32,6 +32,7 @@ module.exports = async function signup(session, username, password, homeserver)
         await nextButton.click();
     }
     //fill out form
+    await session.delay(100);
```
Comment (turt2live) [inline test/end-to-end-tests/src/usecases/signup.js:35]: whenever we `delay` in a test, we generally try to comment on it so we know when we can remove that delay

**Review by turt2live — APPROVED** [review]
otherwise lgtm

**Pull request merged**